### ISSUE TITLE
All reconstructed vectors are the same

### ISSUE THREAD

**Issue by rschuil** [v-v/BiDNN#3]
I'm having trouble reconstructing model1 vectors. The reconstructed values are the same for all inputs.
Here's the command I'm using:
./bidnn.py -n -e 0 -l model.pb.npz -c only-dim2.txt reconstructed.txt 150 2048 1024 1024

(The first 150 values in "only-dim2.txt" are zeros)

Am I doing something wrong..?

**Comment by v-v** [comment]
Maybe there is a bug and it's using the wrong modality (in this case the zeroed input) instead of the correct one. I won't be able to test/debug it before next week, sorry.

The only thing that was properly tested is the multimodal embedding part (the main idea) as we used it in TRECVID this year and it worked well. I tried to make this tool asap but didn't have the opportunity to test everything :/

I'll test it out and try to fix it after a deadline that I currently have, hopefully next week. However if you do find something out before that, feel free to tell me :)

**Comment by v-v** [comment]
Hello again,

I've created a few quick scripts to evaluate crossmodal translation/filling in missing modalities.
This script, say create_data.py, is used to create synthetic data to test with. For the training set, we create the first modality random, while the 2nd modality is just the inverted 1st modality. Each modality is of size 10 and there are a 100 samples. For the test set, in the first 50 samples, we're missing the first modality. The other 50 samples again contain both modalities.
```python
import numpy as np
from sklearn.datasets import load_svmlight_file, dump_svmlight_file

n = 100 # number of samples
d = 10  # each dimension

# create training set
X1_train = np.random.rand(n, d) # 1st modality
X2_train = -X1_train            # 2nd modality is just the inverted 1st
y_train = np.zeros(n)           # unecessary labels
X_train = np.concatenate((X1_train, X2_train), axis=1)

print np.average(np.absolute(X2_train-X1_train)) # check: cca 1.0

# create test set

# first part misses first modality 2nd modality
X2_test1 = np.random.rand(int(n/2), d)   # 2nd modality
X1_test1 = np.zeros_like(X2_test1)       # 1st modality - zeored
X_test1 = np.concatenate((X1_test1, X2_test1), axis=1)

print np.average(np.absolute(X2_test1-X1_test1)) # check: cca 0.5

# second part has both modalities
X1_test2 = np.random.rand(int(n/2), d)   # 1st modality 
X2_test2 = - X1_test2                    # 2nd modality
X_test2 = np.concatenate((X1_test2, X2_test2), axis=1)

print np.average(np.absolute(X2_test2-X1_test2)) # check: cca 1.0

y_test = np.zeros(n)
X_test = np.concatenate((X_test1, X_test2), axis=0)

#saving
dump_svmlight_file(X_train, y_train, "data_train.txt")
dump_svmlight_file(X_test, y_test, "data_test.txt")
```

Next, we train the BiDNN model on the test set (since we have only a 100 samples, we have to also define a smaller batch size).
```
./BiDNN/bidnn.py data_train.txt output_%e.txt 10 10 50 50 -n -e 100 -w 10 -s model_%e.npz -b 10
```

After that, we used the trained model to fill-in the missing modalities of the test set.
```
./BiDNN/bidnn.py data_test.txt output_test.txt 10 10 50 50 -n -l model_final.npz -c -b 10 -e 0
```

Now, we have to files, data_test.txt for the original test set, missing the 1st modality in the first 50 samples and output_test.txt to with that missing part being reconstructed.

The following script loads both files, checks the differences and displays a few reconstructed modalities.
```python
import numpy as np
from sklearn.datasets import load_svmlight_file, dump_svmlight_file

X1, _ = load_svmlight_file("output_test.txt")

# reconstructed file
part1_modality1 = X1[:50, :10].todense()
part1_modality2 = X1[:50, 10:].todense()

part2_modality1 = X1[50:, :10].todense()
part2_modality2 = X1[50:, 10:].todense()

# original file
X2, _ = load_svmlight_file("data_test.txt")

orig1_modality1 = X2[:50, :10].todense()
orig1_modality2 = X2[:50, 10:].todense()

orig2_modality1 = X2[50:, :10].todense()
orig2_modality2 = X2[50:, 10:].todense()

print np.average(np.absolute(part1_modality2 - part1_modality1))
print np.average(np.absolute(part2_modality2 - part2_modality1))

print "\nExamples of reconstructed 1st modality:"

np.set_printoptions(precision=3)
for i in range(5):
    print part1_modality1[i]
```

And we get the following:
```
0.75556461597
1.0171302387

Examples of reconstructed 1st modality:
[[-0.456  0.01  -0.51  -0.438 -0.097 -0.342 -0.024  0.081 -0.113 -0.44 ]]
[[-0.264 -0.509 -0.022 -0.426  0.151 -0.454  0.083 -0.41  -0.137 -0.258]]
[[-0.167 -0.222 -0.404 -0.134 -0.369 -0.276 -0.48  -0.052 -0.233 -0.491]]
[[-0.221 -0.442 -0.465 -0.209 -0.337 -0.037 -0.331 -0.259 -0.451  0.045]]
[[-0.436 -0.008 -0.092 -0.581 -0.225 -0.015 -0.131 -0.435 -0.427 -0.155]]
```
The error of the reconstructed modality is around 0.25, which is not perfect, but it seems to be working and the reconstructed part does not seem to be all the same vector :)

Seems like there is no bug to me. From a scientific standpoint though, have in mind that I'm not sure how good are BiDNNs for filling in missing modalities. It's possible that other architectures would perform better - I never extensively evaluated this feature as the main idea was to perform multimodal embedding, for which you also seem to have good results.

Cheers,
Vedran
